File: src/main/java/tk/hiddenname/smarthome/service/hardware/impl/digital/input/DigitalSensorService.java
package tk.hiddenname.smarthome.service.hardware.impl.digital.input;

import com.pi4j.io.gpio.event.GpioPinListenerDigital;
import tk.hiddenname.smarthome.model.signal.DigitalState;
import tk.hiddenname.smarthome.service.hardware.impl.GPIOService;
import tk.hiddenname.smarthome.service.task.impl.listener.Listener;

public interface DigitalSensorService extends GPIOService {

    DigitalState getState(Long id, boolean reverse);

    GpioPinListenerDigital addListener(Listener listener, Long sensorId, boolean targetSignal,
                                       boolean reverse);

    void removeListener(GpioPinListenerDigital listener, Long sensorId);
}

File: src/main/kotlin/tk/hiddenname/smarthome/service/hardware/impl/digital/input/DigitalSensorService.kt
package tk.hiddenname.smarthome.service.hardware.impl.digital.input

import com.pi4j.io.gpio.event.GpioPinListenerDigital
import tk.hiddenname.smarthome.exception.not_found.SensorNotFoundException
import tk.hiddenname.smarthome.model.signal.DigitalState
import tk.hiddenname.smarthome.service.hardware.impl.GPIOService
import tk.hiddenname.smarthome.service.task.impl.listener.Listener

interface DigitalSensorService : GPIOService {

    @Throws(SensorNotFoundException::class)
    fun getState(id: Long, reverse: Boolean): DigitalState

    @Throws(SensorNotFoundException::class)
    fun addListener(
        listener: Listener, sensorId: Long, targetSignal: Boolean,
        reverse: Boolean
    ): GpioPinListenerDigital

    @Throws(SensorNotFoundException::class)
    fun removeListener(listener: GpioPinListenerDigital, sensorId: Long)
}